### PR TITLE
Implementation of JSON Web Token

### PR DESCRIPTION
**Pull request by semantosoph**
These commits introduce the following things in order to make JSON Web Token (JWT, see https://jwt.io/) usable in takes:

1. A new Signature interface, similar to the Codec interface. The main difference is that while Codecs can encrypt and decrypt data, a Signature can only encrypt (aka sign) data. 
2. An HMAC-SHA implementation of that interface with variations of 256, 384 and 512 bitlength. This is the standard signature algorithm for JWT.
2. A new interface for Token and two implementations that provide a JOSE token (JWT header) and a JWT payload token.
3. A new Pass that accepts JWT from a Bearer header and adds a new JWT to its response (obviously, that works only with JSON responses).

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256 please, pay attention to this pull request

**Comment by 0crat** [comment]
Job `gh:yegor256/takes#743` is in scope.

**Comment by 0crat** [comment]
+15 points just awarded to @semantosoph, total is [+15](http://www.0crat.com/u/semantosoph).

File: src/main/java/org/takes/facets/auth/Token.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+                bytes = new byte[] {};
+            }
+            return bytes;
+        }
+    };
+
+    /**
+     * JSON Web Token.
+     */
+    @SuppressWarnings
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/Token.java]: @semantosoph you can remove that and just initialize all attributes in the constructor, it's a good practice.

File: src/main/java/org/takes/facets/auth/Token.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            return this.jwto;
+        }
+
+        @Override
+        public byte[] encoded() {
+            byte[] bytes;
+            try {
+                bytes = new Base64().encode(this.jwto.toString());
+            } catch (final IOException ignore) {
+                bytes = new byte[] {};
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/Token.java]: @semantosoph this looks a bit weird. Why so? Can't you just throw an unchecked exception here?

File: src/main/java/org/takes/facets/auth/signatures/SiHmac.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    /**
+     * Encrypt the given bytes using HMAC.
+     *
+     * @param bytes
+     *  Bytes to encrypt
+     * @return Encrypted bytes
+     * @throws IOException
+     *  for all unexpected exceptions
+     */
+    @SuppressWarnings("resource")
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/signatures/SiHmac.java]: @semantosoph what's the point of this suppression? Can't you do what Java wants you to do?

**Comment by yegor256** [comment]
@semantosoph looks cool, thanks for your contribution! I made a few minor comments above

File: src/main/java/org/takes/facets/auth/signatures/SiHmac.java (hunk 3 above)
Comment (semantosoph) [inline src/main/java/org/takes/facets/auth/signatures/SiHmac.java]: @yegor256 This is kinda silly. Constructing the Formatter without parameters forces it to use an internal StringBuilder which is totally OK for our use. But that StringBuilder is not closable, so calling close() on the Formatter is useless in terms of saving resources. Eclipse doesn't know this and warns anyway. That's why I suppressed it. I'll close it, if you like it better that way.

File: src/main/java/org/takes/facets/auth/Token.java (hunk 2 above)
Comment (semantosoph) [inline src/main/java/org/takes/facets/auth/Token.java]: @yegor256 Yeah, I don't like declaring `throws` on interface methods. But as of now, it looks like there will always be a reason to throw an IOException, so we can predeclare it anyway.

File: src/main/java/org/takes/facets/auth/Token.java (hunk 1 above)
Comment (semantosoph) [inline src/main/java/org/takes/facets/auth/Token.java]: @yegor256 Yep, will do.

**Comment by semantosoph** [comment]
@yegor256  There you go, I made the requested changes.

I would appreciate it, if you could add me to the contributors list this time.

File: src/main/java/org/takes/facets/auth/PsToken.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                head = nexthead;
+                break;
+            }
+        }
+        final String dot = "\\.";
+        if (head.length() > 0) {
+            final String jwt = head.trim().split(" ", 2)[1].trim();
+            final byte[] jwtheader = jwt.split(dot)[0].getBytes();
+            final byte[] jwtpayload = jwt.split(dot)[1].getBytes();
+            final byte[] jwtsign = jwt.split(dot)[2].getBytes();
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/PsToken.java]: @semantosoph this code looks a bit weird. Why can't you do:

```
final String[] parts = jwt.split(dot);
final byte[] jwtheader = parts[0].getBytes();
final byte[] jwtpayload = parts[1].getBytes();
final byte[] jwtsign = parts[2].getBytes();
```

File: src/main/java/org/takes/facets/auth/signatures/SiHmac.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+     * Encrypt the given bytes using HMAC.
+     *
+     * @param bytes
+     *  Bytes to encrypt
+     * @return Encrypted bytes
+     * @throws IOException
+     *  for all unexpected exceptions
+     */
+    private byte[] encrypt(final byte[] bytes) throws IOException {
+        final byte[] encrypted;
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/signatures/SiHmac.java]: @semantosoph why do you need this variable? Can't you just do 

```
return formatter.toString().getBytes();
```

**Comment by yegor256** [comment]
@semantosoph a few more comments above. Of course, we can add you to the list, but we don't have that list now :) Feel free to add it to README, just like it's done in github.com/yegor256/cactoos

File: src/main/java/org/takes/facets/auth/PsToken.java (hunk 4 above)
Comment (semantosoph) [inline src/main/java/org/takes/facets/auth/PsToken.java]: Can do.

File: src/main/java/org/takes/facets/auth/signatures/SiHmac.java (hunk 5 above)
Comment (semantosoph) [inline src/main/java/org/takes/facets/auth/signatures/SiHmac.java]: Indeed, I can.

**Comment by semantosoph** [comment]
@yegor256 Changes done. I'll compile the contributors list once I've got a little spare time.

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@semantosoph when you want me to release a new version, say it here, not in Twitter :)

**Comment by yegor256** [comment]
@rultor release, tag is `1.6`

**Comment by 0crat** [comment]
Job `gh:yegor256/takes#743` assigned to @caarlos0, please go ahead ([policy](http://datum.zerocracy.com/pages/policy.html)).

**Comment by 0crat** [comment]
@caarlos0 this job was assigned to you 8 days ago.
It will be taken away from you after 10 days from start (this is our policy).

**Comment by 0crat** [comment]
Oops! Job `gh:yegor256/takes#743` is not assigned to anyone.

**Comment by 0crat** [comment]
Oops! Job `gh:yegor256/takes#743` is not assigned to anyone.

**Comment by 0crat** [comment]
Oops! Job `gh:yegor256/takes#743` is not assigned to anyone.

**Pull request merged**